src/hotspot/os/linux/os_linux.cpp

```diff
@@ -3918,139 +3918,98 @@ static void warn_on_large_pages_failure(char* req_addr, size_t bytes,
-  char* lp_start = align_up(start, large_page_size);
-  char* lp_end   = align_down(end, large_page_size);
-
-  size_t lp_bytes = lp_end - lp_start;
-
-  assert(is_aligned(lp_bytes, large_page_size), "Must be");
+  assert(is_aligned(req_addr, os::large_page_size()), "Must be");
+  assert(is_aligned(alignment, os::vm_allocation_granularity()), "Must be");
+  assert(is_power_of_2(os::large_page_size()), "Must be");
+  assert(bytes >= os::large_page_size(), "Shouldn't allocate large pages for small sizes");
 
-  if (lp_bytes == 0) {
-    // The mapped region doesn't even span the start and the end of a large page.
-    // Fall back to allocate a non-special area.
-    ::munmap(start, end - start);
+  // We only end up here when at least 1 large page can be used.
+  // If the size is not a multiple of the large page size, we
+  // will mix the type of pages used, but in a decending order.
+  // Start of by reserving a range of the given size that is
+  // properly aligned. If a requested address is given it will
+  // be used and it must be aligned to both the large page size
+  // and the given alignment. The larger of the two will be used.
+  size_t required_alignment = MAX(os::large_page_size(), alignment);
+  char* const aligned_start = anon_mmap_aligned(req_addr, bytes, required_alignment);
+  if (aligned_start == NULL) {
     return NULL;
   }
 
-  int prot = exec ? PROT_READ|PROT_WRITE|PROT_EXEC : PROT_READ|PROT_WRITE;
-  int flags = MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED;
-  void* result;
+  // Start of by mapping large pages.
+  size_t large_bytes = align_down(bytes, os::large_page_size());
+  char* large_mapping = reserve_and_commit_special(large_bytes, os::large_page_size(), aligned_start, exec);
 
-  // Commit small-paged leading area.
-  if (start != lp_start) {
-    result = ::mmap(start, lp_start - start, prot, flags, -1, 0);
-    if (result == MAP_FAILED) {
-      ::munmap(lp_start, end - lp_start);
-      return NULL;
-    }
+  if (bytes == large_bytes) {
+    // The size was large page aligned so return the large mapping.
+    return large_mapping;
   }
 
-  // Commit large-paged area.
-  flags |= MAP_HUGETLB | hugetlbfs_page_size_flag(os::large_page_size());
-
-  result = ::mmap(lp_start, lp_bytes, prot, flags, -1, 0);
-  if (result == MAP_FAILED) {
-    warn_on_large_pages_failure(lp_start, lp_bytes, errno);
-    // If the mmap above fails, the large pages region will be unmapped and we
-    // have regions before and after with small pages. Release these regions.
-    //
-    // |  mapped  |  unmapped  |  mapped  |
-    // ^          ^            ^          ^
-    // start      lp_start     lp_end     end
-    //
-    ::munmap(start, lp_start - start);
-    ::munmap(lp_end, end - lp_end);
+  // The requested size requires some small pages as well.
+  char* small_start = aligned_start + large_bytes;
+  size_t small_size = bytes - large_bytes;
+  if (large_mapping == NULL) {
```

Comment:
See comment above

Reply:
I do the check here because I need the `small_start` and `small_size` values if `large_mapping == NULL` and there was additional memory reserved for small pages.